File: hw2/task3/Monster.py
"""
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Chao Yu
 * Student ID : 1155053722
 * Email Addr : ychao5@cse.cuhk.edu.hk
"""

from Pos import Pos

class Monster():
    def __init__(self, monster_id, health_capacity):
        self.monster_id = monster_id
        self.health_capacity = health_capacity
        self.health = health_capacity
        self.pos = Pos()
        self.drop_item_list = []
        self.hint_list = []

    def add_drop_item(self, key):
        self.drop_item_list.append(key)
    
    def add_hint(self, monster_id):
        self.hint_list.append(monster_id)

    def get_monster_id(self):
        return self.monster_id

    def get_pos(self):
        return self.pos

    def set_pos(self, row, column):
        self.pos.set_pos(row, column)

    def get_health_capacity(self):
        return self.health_capacity

    def get_health(self):
        return self.health

    def lose_health(self):
        self.health -= 10
        return self.health <= 0

    def recover(self, healing_power):
        self.health = healing_power

    def action_on_soldier(self, soldier):
        if self.health <= 0:
            self.talk("You had defeated me.\n\n")
        else:
            if self.require_key(soldier.get_keys()):
                self.fight(soldier)
            else:
                self.display_hints()
    
    def require_key(self, keys):
        return self.monster_id in keys
    
    def display_hints(self):
        self.talk("Defeat Monster " + str(self.hint_list) + " first.\n\n")

    def fight(self, soldier):
        fight_enabled = True
        
        while fight_enabled:
            print("       | Monster{} | Soldier |".format(self.monster_id))
            print("Health | %8d | %7d |\n" % (self.health, soldier.get_health()))
            choice = input("=> What is the next step? (1 = Attack, 2 = Escape, 3 = Use Elixir.) Input: ")

            if choice == "1":
                if self.lose_health():
                    print("=> You defeated Monster{}.\n".format(self.monster_id))
                    self.drop_items(soldier)
                    fight_enabled = False
                else:
                    if soldier.lose_health():
                        self.recover(self.health_capacity)
                        fight_enabled = False
            elif choice == "2":
                self.recover(self.health_capacity)
                fight_enabled = False
            elif choice == "3":
                if soldier.get_num_elixirs() == 0:
                    print("=> You have run out of elixirs.\n")
                else:
                    soldier.use_elixir()
            else:
                print("=> Illegal choice!\n")

    def drop_items(self, soldier):
        for item in self.drop_item_list:
            soldier.add_key(item)
    
    def talk(self, text):
        print("Monster{}: ".format(self.monster_id) + text, end="")
    
    def display_symbol(self):
        print("M", end="")

File: hw2/task4/java/Monster.java
/*
* CSCI3180 Principles of Programming Languages
*
* --- Declaration ---
*
* I declare that the assignment here submitted is original except for source
* material explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
*
* Assignment 2
* Name : Chao Yu
* Student ID : 1155053722
* Email Addr : ychao5@cse.cuhk.edu.hk
*/

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Scanner;

public class Monster {
  private int monsterID;
  private int healthCapacity;
  private int health;
  private Pos pos;
  private ArrayList<Integer> dropItemList;
  private ArrayList<Integer> hintList;

  public Monster(int monsterID, int healthCapacity) {
    this.monsterID = monsterID;
    this.healthCapacity = healthCapacity;
    this.health = healthCapacity;
    this.pos = new Pos();
    this.dropItemList = new ArrayList<Integer>();
    this.hintList = new ArrayList<Integer>();
  }

  public void addDropItem(int key) {
    this.dropItemList.add(key);
  }

  public void addHint(int monsterID) {
    this.hintList.add(monsterID);
  }

  public int getMonsterID() {
    return this.monsterID;
  }

  public Pos getPos() {
    return this.pos;
  }

  public void setPos(int row, int column) {
    this.pos.setPos(row, column);
  }

  public int getHealthCapacity() {
    return this.healthCapacity;
  }

  public int getHealth() {
    return this.health;
  }

  public boolean loseHealth() {
    this.health -= 10;
    return this.health <= 0;
  }

  public void recover(int healingPower) {
    this.health = healingPower;
  }

  public void actionOnSoldier(Soldier soldier) {
    if (this.health <= 0) {
      this.talk("You had defeated me.%n%n");
    } else {
      if (this.requireKey(soldier.getKeys())) {
        this.fight(soldier);
      } else {
        this.displayHints();
      }
    }
  }

  public boolean requireKey(HashSet<Integer> keys) {
    return keys.contains(this.monsterID);
  }

  public void displayHints() {
    this.talk("Defeat Monster " + this.hintList + " first.%n%n");
  }

  public void fight(Soldier soldier) {
    boolean fightEnabled = true;

    while (fightEnabled) {
      System.out.printf("       | Monster%d | Soldier |%n", this.monsterID);
      System.out.printf("Health | %8d | %7d |%n%n", this.health, soldier.getHealth());
      System.out.printf("=> What is the next step? (1 = Attack, 2 = Escape, 3 = Use Elixir.) Input: ");

      Scanner sc = new Scanner(System.in);

      String choice = sc.nextLine();

      if (choice.equalsIgnoreCase("1")) {
        if (this.loseHealth()) {
          System.out.printf("=> You defeated Monster%d.%n%n", this.monsterID);
          this.dropItems(soldier);
          fightEnabled = false;
        } else {
          if (soldier.loseHealth()) {
            this.recover(this.healthCapacity);
            fightEnabled = false;
          }
        }
      } else if (choice.equalsIgnoreCase("2")) {
        this.recover(this.healthCapacity);
        fightEnabled = false;
      } else if (choice.equalsIgnoreCase("3")) {
        if (soldier.getNumElixirs() == 0) {
          System.out.printf("=> You have run out of elixirs.%n%n");
        } else {
          soldier.useElixir();
        }
      } else {
        System.out.printf("=> Illegal choice!%n%n");
      }
    }
  }

  public void dropItems(Soldier soldier) {
    for (Integer item : this.dropItemList) {
      soldier.addKey(item);
    }
  }

  public void talk(String text) {
    System.out.printf("Monster%d: " + text, this.monsterID);
  }

  public void displaySymbol() {
    System.out.printf("M");
  }
}